From 2a46880f847e6f2c82cef0828bb2edd934b445b4 Mon Sep 17 00:00:00 2001
From: Robby Cai <robby.cai@nxp.com>
Date: Wed, 18 Apr 2018 17:16:47 +0800
Subject: [PATCH 3692/5242] MLK-18095 media: v4l2: Add VIDIOC_EXPBUF ioctl for
 DMABUF mode

commit  7ce6267a9e8bc883a25eca9f392810cdf1948d8c from
https://source.codeaurora.org/external/imx/linux-imx.git

Add VIDIOC_EXPBUF ioctl which exports a buffer as a DMABUF file descriptor.
After calling VIDIOC_EXPBUF the fd field will be set by a driver. This is a
DMABUF file descriptor. The application may pass it to other DMABUF-aware
devices. See V4L2 spec for more details.

Signed-off-by: Robby Cai <robby.cai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/mxc/capture/mx6s_capture.c |   12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/drivers/media/platform/mxc/capture/mx6s_capture.c b/drivers/media/platform/mxc/capture/mx6s_capture.c
index 00acfd4..c84089c 100644
--- a/drivers/media/platform/mxc/capture/mx6s_capture.c
+++ b/drivers/media/platform/mxc/capture/mx6s_capture.c
@@ -1490,6 +1490,17 @@ static int mx6s_vidioc_querycap(struct file *file, void  *priv,
 	return 0;
 }
 
+static int mx6s_vidioc_expbuf(struct file *file, void *priv,
+			     struct v4l2_exportbuffer *eb)
+{
+	struct mx6s_csi_dev *csi_dev = video_drvdata(file);
+
+	if (eb->type == V4L2_BUF_TYPE_VIDEO_CAPTURE)
+		return vb2_expbuf(&csi_dev->vb2_vidq, eb);
+
+	return -EINVAL;
+}
+
 static int mx6s_vidioc_streamon(struct file *file, void *priv,
 			       enum v4l2_buf_type i)
 {
@@ -1672,6 +1683,7 @@ static int mx6s_vidioc_enum_frameintervals(struct file *file, void *priv,
 	.vidioc_enum_input    = mx6s_vidioc_enum_input,
 	.vidioc_g_input       = mx6s_vidioc_g_input,
 	.vidioc_s_input       = mx6s_vidioc_s_input,
+	.vidioc_expbuf        = mx6s_vidioc_expbuf,
 	.vidioc_streamon      = mx6s_vidioc_streamon,
 	.vidioc_streamoff     = mx6s_vidioc_streamoff,
 	.vidioc_g_parm        = mx6s_vidioc_g_parm,
-- 
1.7.9.5

